### ISSUE TITLE
Support click-through of transparency

### ISSUE THREAD

**Issue by tommoor** [atom/atom-shell#1335]
As commented here, https://github.com/atom/atom-shell/pull/949#issuecomment-87087841 - transparent windows unfortunately are of limited use unless you can click through the transparent area.

I noticed that in the documentation it is stated that this is blocked by an upstream bug. The nw.js project has [support for this](https://github.com/nwjs/nw.js/wiki/Transparency#click-through-transparency), but i'm not sure how different their implementation is...

I think this is worth the issue to track discussion and implementation, so here it is.

**Comment by thomasjo** [comment]
While nw.js does support it, it comes with some pretty hefty drawbacks. First of all it requires disabling GPU acceleration, and secondly it only works on OS X and Windows (based on their documentation).

**Comment by CharlieHess** [comment]
+1, this one's plaguing Slack too, since we display all of our notifications in a transparent window.

**Comment by YourCyborg** [comment]
+1, this is critical for the project I'm working on.

**Comment by MarcFowler** [comment]
+1, this would be excellent.

**Comment by geelen** [comment]
Interested in hearing the latest on this. For my purposes, I'd love if we could set a transparent window to have *no* click area. So you could have an always-on-top full screen overlay that didn't catch clicks.

**Comment by kitak** [comment]
:+1:

**Comment by joshwnj** [comment]
:+1:

**Comment by sevengo8378** [comment]
+1

**Comment by TimvdEijnden** [comment]
I'm interested too, i'm building an app with a semi transparent window that's on top of other applications but mouseevents need to get through to the other applications.

**Comment by mpouleijn** [comment]
+1

**Comment by wvandenborne608** [comment]
+1

**Comment by bloemendal** [comment]
+1

**Comment by ralfschimmel** [comment]
+1

**Comment by johanvdlaak** [comment]
+1

**Comment by diimdeep** [comment]
+1

**Comment by yohaia** [comment]
+1

Any news about it?

**Comment by lukeapage** [comment]
the linked chromium bug is marked wont-fix
https://code.google.com/p/chromium/issues/detail?id=387234

the last comment is:
 > #11 NekR.Fab...@gmail.com
 > You know, just btw, people in Atom team can do it by themselves if their users need it, right?

so nothing is currently going to happen until someone on the atom team decides to do it or persuades the chrome team it is worth putting on their backlog (and they said it was low priority).

**Comment by molster** [comment]
Just ran into this today, It has put a -massive- issues into my plans of using a new window has an overlay. (in fact, it is now impossible)

I would love a fix for this!

**Comment by Schrims** [comment]
+1

**Comment by deibeljc** [comment]
+100

**Comment by devployment** [comment]
Is there any work in progress for this? Will there be a chance for the Electron team to make this possible without being added to Chromium?

**Comment by Alisamix** [comment]
+1

**Comment by aestrro** [comment]
+1

**Comment by tex0l** [comment]
+1

**Comment by arantes555** [comment]
+1

**Comment by stanier** [comment]
+1

Assuming this won't be resolved through the Chromium code, is there a plausible way to have the application listen for a click event and then trigger the proper counterpart in GTK or similar?

**Comment by kutu** [comment]
i use ahkscript
```
#c::
    WinSet, ExStyle, ^0x20, A
    WinSet, Transparent, 255, A
    return
```
press Win+C on active window, and its click-through now

**Comment by b1015062Furukawa** [comment]
+1

**Comment by sergey-spivakov** [comment]
+1

**Comment by Dafrok** [comment]
+1

**Comment by blade254353074** [comment]
+!

**Comment by xeoneux** [comment]
@zcbenz Any update on this? 😄

**Comment by matanshapira** [comment]
+1

**Comment by bobasoft** [comment]
+1

**Comment by TekSiDoT** [comment]
+1

**Comment by zcbenz** [comment]
To people watching this issue: if you simply want to make the whole window click-though, you can use the new [win.setIgnoreMouseEvents](http://electron.atom.io/docs/api/browser-window/#winsetignoremouseeventsignore) API introduced in v1.2.2.

**Comment by zcbenz** [comment]
Also on making specified regions transparent, I have come up with some ideas on the implementation (but do not assume it would be done in near future):

For Linux, it is actually very easy by using the `XShapeCombineRectangles` API, the implementation of `SetIgnoreMouseEvents` is just setting only a region of (0, 0, 1, 1) to accept mouse events.

For Windows, it is easy to set the shape of window, but there is no API to control which regions can accept mouse events.

For OS X, either setting window shape or setting click-though regions is impossible if you don't have control to the rendering code.

And since Chromium does complex hardware acceleration, it becomes super hard to implement custom click-through regions. NW.js does it by disabling hardware acceleration, and patching the rendering code, this is an approach that I definitely don't want to take.

My idea for the implementation on OS X/Windows is, we just make the whole window click-through, and then create a transparent window as child window, and use it to capture all mouse events. Since we have full control of the child window, we can set it to the shape of the regions that users do not want to be click-though, and then send all mouse and keyboard events received in the child window to the parent window.

In this way we don't have to patch Chromium, so it is easy to maintain and hardware acceleration works. The downside is we have to very carefully redirect mouse and keyboard events, to make the window behave as if it there is no child window.

**Comment by StickyCube** [comment]
If, like me, your use case for this feature is fairly simple [this gist](https://gist.github.com/StickyCube/ed79421bc53cba38f5b74b060d3f15fa) might be of interest. It is a basic implementation of what @zcbenz is describing, creating one clickable area which redirects mouse events to the parent.

**Comment by cksachdev** [comment]
+1

**Comment by ijse** [comment]
+1

**Comment by soupman99** [comment]
+1

**Comment by Schottii** [comment]
+1

**Comment by ButzYung** [comment]
During the development of my Windows gadget project which requires click-thru, I have unintentionally come up with a somewhat fancy but practical solution which works to define a clickcable area of truly arbitrary shape.

In the renderer process, use "screen.getCursorScreenPoint()" to keep tracking the cursor position, and then send the coordinates to the main process regularly. On the main process, use "webContents.capturePage()" to capture the image of that exact 1x1 pixel on the renderer window. Since it's just 1 pixel, the performance cost is ignorable, and then you can define whether that pixel is clickable or not by looking at the opacity value. This way, you can practically define a truly pixel-correct clickable area. The only issue left here is a function to detect the mousedown state regardless of the cursor position.

**Comment by amhoho** [comment]
Electron click through transparency example
https://gist.github.com/StickyCube/ed79421bc53cba38f5b74b060d3f15fa

**Comment by varavut** [comment]
@ButzYung Could you please tell me that how can I get opacity value from NativeImage (webContents.capturePage() return NativeImage)? Thanks.

**Comment by ButzYung** [comment]
@varavut Capturing a 1x1 image as an example, it would be something like this.

  ```
webContents.capturePage({x:0,y:0,width:1,height:1}, function(image) {
    var buffer = image.getBitmap()

    // send the data back to renderer for processing, with buffer[3] being the alpha channel
    webContents.send('capturePage', 'RESULT|' + [buffer[0],buffer[1],buffer[2],buffer[3]])
  });
```

**Comment by cecilpeng** [comment]
@ButzYung Thanks for your solution. I tried it, the key is the interval of tracking. 100ms for my computer, looks well. If I used `webContents.capturePage`, CPU cost high. I would define the area which is click-through, just check the point if in the area.

**Comment by ButzYung** [comment]
@cecilpeng In my case I give a 200ms delay after the current webContents.capturePage has finished, instead of doing another capture immediately. That's roughly 4-5 captures/second, enough for the purpose of checking transparency, and the CPU usage is ignorable.

**Comment by cecilpeng** [comment]
@ButzYung yes, it is better to do another capture after last has finished. I'm still worried about the timeout solution, it's not smart enough. We can't know what happens on the users' computer. I recommend that javascript is a good languague to observe instead of inspect.

**Comment by sancarn** [comment]
I'm very much a HTML noob but is it not possible to simply:

1. setup a 'onMouseLeave' event on HTML elements which you want to be clickable.
>>> this mouse leave event executes win.setIgnoreEvents(true)
2. setup a 'onMouseEnter' event on HTML elements which you want to be clickable.
>>> this mouse enter event executes win.setIgnoreEvents(false)

Due to my lack of HTML knowledge there may be an issue with this in certain scenarios... But in my head it seems to be the, currently, ideal workaround. Except for the extra setup time required of course?

**Comment by ButzYung** [comment]
@sancarn The problem with your approach is that when .setIgnoreMouseEvents is set to true in step 1, there is no more mouse event and thus the usual HTML-driven "onMouseOver" event in step 2 will not fire anymore.

But hey, now that I think about it, if your approach is combined with my "capturePage" approach, which can detect mouseover on/off by comparing opacity without the usual HTML events, it may produce interesting results. I have actually tried, and guess what, it works lol. Now, you can have a truly pixel-correct clickable area.

**Comment by sancarn** [comment]
Lol riiight. Hilarious that I didn't see that as a flaw xD

**Comment by sancarn** [comment]
When it comes to getting the pixel, I haven't tested this but on Windows it may be easier to use [hookWindowMessages()](https://github.com/electron/electron/blob/master/docs/api/browser-window.md#winhookwindowmessagemessage-callback-windows) with message [WM_MOUSEMOVE](https://wiki.winehq.org/List_Of_Windows_Messages) = 512.

**Comment by stanleyxu2005** [comment]
When I set `transparent=true` I cannot double click to maximize the window. The only reason I need to set transparency is create a rounded corner window. If this issue is resolved, we can definitively resolve many issues and save many efforts. Really looking forward for someone to implement this.

**Comment by DaniGuardiola** [comment]
Hi, I just wanted to +1 this bug because it would be a really nice addition for my use case. This would really make a difference in our UX. A huge one.

_Sent from my Google Pixel XL using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by andens** [comment]
If you're on Windows with a similar use case to mine you may find my recently accepted PR useful. It adds a parameter to `setIgnoreMouseEvents` that makes the window click-through as usual, but still allows you to listen DOM events such as `mouseleave`. Provided that you can determine when the window should be click-through based on mouse movements you should be able to opt out and return to normal as in [this example](https://github.com/andens/electron/blob/d4bbd7c7abe315e6aa4f62ae57d130f6dde4b531/docs/api/frameless-window.md#forwarding). For a bit more context, my use case was that I wanted a hole in the web page (possibly covered by other elements). The example shows how I enabled/disabled click-through as I entered/left an element. The transparency itself was made using a backdrop filter via experimental CSS features.

Unfortunately I don't know Linux or Mac GUI programming so I'm unaware whether a similar approach is possible there. Still, I figured that I'll inform of it in case somebody finds it useful.

**Comment by DaniGuardiola** [comment]
@andens that sounds great really! Please someone implement the same for Linux and Mac! That would be good enough for me!

_Sent from my Google Pixel XL using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by edunad** [comment]
@andens i seem to be having a weird issue with it, in order to get it working correctly i need to pass forward on both false and true, else it won't work. After doing that, i need to reload the the webcontent at least once. After all this, seems to be working flawlessly :)

Doesn't work (follows example)
`
      divThing.addEventListener('mouseleave', () => {
        win.setIgnoreMouseEvents(true, {forward: true});
      });

      divThing.addEventListener('mouseenter', () => {
        win.setIgnoreMouseEvents(false);
      });
`

Works after reloading at least once
`
      divThing.addEventListener('mouseleave', () => {
        win.setIgnoreMouseEvents(true, {forward: true});
      });

      divThing.addEventListener('mouseenter', () => {
        win.setIgnoreMouseEvents(false, {forward: true});
      });
`

**Comment by andens** [comment]
@edunad Sorry for the delay; I'm swamped with other stuff at the moment. If I remember correctly, forward should make no difference whatsoever when the first parameter is false, but I may be mistaken. The code you posted forwards when leaving the div and stops when entering, whereas the documentation example does it the other way around. Is this intended? Even so, the semantics should be the same, just that the forwarding area is essentially inverted, but perhaps there are some corner cases in how Chromium behaves.

One quick thing I can think of is that you probably don't get the leave event (and thus begin forwarding) until you have first entered the div. Have you considered this? Hopefully I can find some time to try it out.

**Comment by edunad** [comment]
@andens No problem! Take all the time you need, i'm in no hurry. (ended up revering back to 1.7 + using disableHardwareAcc() )

What i needed was the opposite of the example (ignore inputs if not hovering the div).
I did try the example and i had the same issue :(

Basically it's a small div on each the sides of the screen,
![](https://i.imgur.com/0aqgi9L.png)
I've added console logs to it, and it seems mouseenter / mouseleave would get triggered, but once mouseleave was triggered i coulnd't re-trigger the mouseenter to disable ignoreMouse.

Would be nice to use hardware acceleration and click electron elements if they are not completely invisible (opacity 0)

**Comment by main** [comment]
+1, this is critical for the project I'm working on.

**Comment by AlexCatch** [comment]
A great solution that works flawlessly for me is the answer by @ButzYung, rather than checking opacity, I simply check if the mouse is within my button container div, if it's inside I enable mouse events, if it's outside I disable them and the window becomes click through again.

**Comment by DaniGuardiola** [comment]
Do we have any Linux / Mac programmers who could try and implement the same
thing that the PR Andreas is mentioning does in Windows? I would really
really love to see that working <3

**Comment by paulius005** [comment]
Does anyone have a workaround for mac for this?

**Comment by JokerQyou** [comment]
`window.setIgnoreMouseEvents(true)` actually works on macOS.

**Comment by andens** [comment]
@paulius005 It's only implemented on Windows because it uses the native WinAPI. This kind of functionality is highly dependent on OS features and (I would guess that) it's probably very difficult to realize unless the OS provides explicit support.

What you suggest sound very similar to what is done for Windows. If you know low-level macOS programming, feel free to study https://github.com/electron/electron/pull/10183 for inspiration and I'll assist with the ideas as best as I can.

**Comment by leidsongermano** [comment]
Any news about this feature? I'm trying to use setIgnoreMouseEvents on Windows 10 app without success. I'm also using the app with transparent option enabled.

**Comment by edunad** [comment]
@leidsongermano i'm currently using v1.8.7 with ```app.disableHardwareAcceleration();```,  seems to work fine. The only downside is disabling the hardware acceleration

**Comment by leidsongermano** [comment]
@edunad i was testing with the electron-quickstart which uses electron 0.36.*. After that i updated to the latest versionp (2.0.2) and it works pretty fine now. Thanks!

**Comment by adeperio** [comment]
@edunad that worked for me. disableHardwareAcceleration slowed my app down too much on Mac though but for some reason was fine on Windows. So ended up disableHardwareAcceleration just for Windows machines

**Comment by edunad** [comment]
@singuerinc yes, if you enable ```disableHardwareAcceleration``` and set the app transparency to true, only solid and ```if you set the opacity to 0.1``` will be clickable. Do take note about the slowdown on macs (check @adeperio's response)

**Comment by singuerinc** [comment]
@edunad I just tried a bunch of different ways, but it doesn't work.
I set up a empty-transparent-fullsize window without any element and I'm not able to click-through, I have no idea what I'm doing wrong.
`window.setIgnoreMouseEvents(true)` is the only way, but as I said my app became useless.

**Comment by edunad** [comment]
Try using an older version of electron, i'm currently using ```1.8.7```

**Comment by toonvanvr** [comment]
I've written a simple workaround class where you can choose whitelist or blacklist functionality for clicking through HTMLElements. It's available as [electron-transparency-mouse-fix](https://github.com/toonvanvr/electron-transparency-mouse-fix) and also works for rounded corners! It was written from scratch today, so there still may be some issues.

Let's hope some of you can use this.

**Comment by etisdew** [comment]
I have been grepping for the actual implementation of setIgnoreMouseEvents and I'm coming up dry besides finding the interface for its options. Getting a bit frustrated, could I tack this in for an explanation of where its behavior is actually defined?

**Comment by nornagon** [comment]
If you're using Electron >3.0.0 (currently in beta) and you only need totally-transparent areas to be click-through-able, try the setWindowShape API introduced in #13789.

**Comment by singuerinc** [comment]
Looks like is not available on Mac.

**Comment by andens** [comment]
@etisdew I believe what you're looking for can be found in `atom/browser/native_window_views.cc` (Windows/Linux) and in `atom/browser/native_window_mac.mm` (Mac).

**Comment by toonvanvr** [comment]
Some people in this thread already proposed adding click-through based on transparency using the GPU, but would that really be useful? Most windows need a shadow to make window hierarchy clear. I've been working on an always-on-top floating widget/helper tool which really needs the shadows to stand out. Since these have an alpha>0, click through would be impossible. After working some more on the [library](https://github.com/toonvanvr/electron-transparency-mouse-fix) posted above, I found out that most of the language features already exist to support native click-through.

Would it be possible to implement the way this library works natively?

### JS/CSS equivalent
```CSS
body {pointer-events: none}
```
```JS
window.addEventListener('mousemove', event =>
  if (event.target === document.documentElement) // <html>-element
    setIgnoreMouseEvents(true, {forward: true)   // {forward: true} keeps generating MouseEvents
  else
    setIgnoreMouseEvents(false)
)
```
The `pointer-events` CSS rule is standardized and prevents elements from being the `event.target`. You can add something like `{transparent: true, clickThrough: 'pointer-events'}` to the `BrowserWindow` options to activate this feature.

There's more to it when actually implementing it and supporting drag-events etc, but it doesn't seem to cause anything near a bottleneck. `MouseEvent`s are generated anyhow, so performance shouldn't really be impacted. The ignoreMouseEvents state can be cached so only mouseenter/mouseleave triggers native window code.


![animated png recorded@60fps, played@60fps](https://user-images.githubusercontent.com/2529712/47581685-6e8bc700-d952-11e8-843a-bb4a3c2309e2.png)

**Comment by metaa** [comment]
Any news here? 🙃

**Comment by singuerinc** [comment]
@metaa I partially fixed it in Mac (looks more like a hack) with `setIgnoreMouseEvents` (Electron 3)

https://github.com/singuerinc/overlay/blob/master/app/components/helpers/mouseEvents.ts#L3

So basically every time you go over an element in your app you have to `setIgnoreMouseEvents(false)` and when you go out the opposite. So it's quite expensive depending on your app. I'd say that it works quite well, although sometimes it's really hard to debug since the entire Electron becomes "non-clickable", in that case you can't use the devtools.

**Comment by toonvanvr** [comment]
MacOs with the current version of Electron doesn't seem to have the problems I encountered months ago on Windows / Linux. 🙌

This is a minimal setup which works out of the box for me:

**javascript:**
```javascript
const setIgnoreMouseEvents = require('electron').remote.getCurrentWindow().setIgnoreMouseEvents

window.addEventListener('pointerover', function mousePolicy (event) {
  mousePolicy._canClick = event.target === document.documentElement ?
    mousePolicy._canClick && setIgnoreMouseEvents(true, {forward: true}) :
    mousePolicy._canClick || setIgnoreMouseEvents(false) || 1
})
```

**css:**
```css
body {pointer-events: none}
.clickable {pointer-events: auto}
```

I really love the way this uses native CSS functionality. You can make an item click-through on the fly within devtools, the code footprint is minimal and it's pretty good performance wise too. The way inheritance works in CSS also suits this use case.

**Comment by mynetx** [comment]
@toonvanvr Does this work on any macOS version, provided that we use the latest Electron version, as you said?

**Comment by toonvanvr** [comment]
@mynetx Unfortunately, I have no idea about that since macOS 10.14 Mojave is the only version I ever used. The key feature that needs to be supported is `{forward: true}` while the rest should be platform-independent.

**Comment by david-winder-kaltura** [comment]
I am using electron 4.0.0 with MacOs 10.14.5 and even if I used the `{forward: true}` I still cannot get the mouse event (even I listen on the `mouseover` events). 
I used: `win.setIgnoreMouseEvents(true, { forward: true});` but then all mouse event are blocked.
On win8.1 with electron 4.0.0 it is working.
@toonvanvr - should it work? am I missing anything?

**Comment by toonvanvr** [comment]
@david-winder-kaltura
I haven't really been been using Electron anymore for a while, but last time I checked, it worked great without opening the devTools. Once you open the console, it's a mess with race conditions etc. to keep the mouse events and transparency in check.

**Comment by LauraAubin** [comment]
I was inspired by [this](https://discuss.atom.io/t/handling-click-events-in-the-dom/41642) solution to do the following in my project:

### From main, 

Listen for replies to the emitted `"hideWindow"`:

```js
const { ipcMain } = require('electron');

ipcMain.on('hideWindow', () => {
  mainWindow.hide();
});
```

### From your application,

If an element with the id `mainElement` is clicked, send a response to `"hideWindow"`:

```ts
const { ipcRenderer } = window.require('electron');

// so that we can write window.require
declare global {
  interface Window {
    require: any;
  }
}

...

export function hideWindow(event: any) {
  let hideWindow = true;

  // some looping so that we can check for the id on all parent/child elements
  event.path.map(object => {
    if (object.id == 'mainElement') {
      hideWindow = false;
    }
  });

  hideWindow && ipcRenderer.send('hideWindow');
}

document.addEventListener('mousedown', hideWindow);
```

### A downside

- You need to add the id `"mainElement"` to all of your separate components. This may be annoying to maintain if you're rendering multiple separate elements.

### What to takeaway from this

I found this to be a clean solution for an application that works similar to [Spotlight](https://support.apple.com/en-ca/HT204014). I can show the application using keyboard commands, and hide it if I ever click outside of actual DOM elements.

**Comment by DEDaniel** [comment]
Have you all tried to set

`app.commandLine.appendSwitch( 'disable-gpu-compositing' );`

before the app ready event?

**Comment by vifird** [comment]
@DEDaniel Does it works?

**Comment by WiIdcard** [comment]
@LauraAubin's solution works fairly well, I'm using it myself.

Another downside though, is the clicks don't actually propagate through the transparent area. So if the user is trying to click something below your window, they will have to click twice to actually accomplish it.

**Comment by DEDaniel** [comment]
@vifird yes for me this is working on Windows 7 and 10 using Electron version 6.0.2.
I came upon this issue because I wanted to disable an area in my css grid and having it transparent.
After having the transparent area I realised that the underlying application didn't received any click events. Therefore I added the mentioned line in my main.js Electron file and the problem is solved. 🥳

**Comment by codemanki** [comment]
Could anybody tell me if there is a way to capture the coordinates of the click through event? 

I want to have a transparent overlay that is clickable through ( I achieved this using solutions from this thread ), but also want to capture the coordinates of those click throughs ( no clue on how to do this) . I would appreciate if anybody could tell me if this is possible and maybe point me toward the solution.
(Feeling a bit desperate, already spent multiple hours on trying to achieve this, but no result :( )

**Comment by LauraAubin** [comment]
> Could anybody tell me if there is a way to capture the coordinates of the click through event?

@codemanki Hey!

Given that your overlay expands the whole area that you want to capture, you should be able to get the coordinates of the mouse using some JS with the `event` handler such as `onClick` or even `onMouseMove`.

I whipped up a little Codepen demonstrating this, though there are multiple approaches you can take.

https://codepen.io/LauraAubin/pen/NWPKrrq

Also note that I believe this solution will only work if the transparent overlay covers the entire space your mouse will hover over. That is, this won't work if you click outside of the Electron app.

**Comment by codemanki** [comment]
@LauraAubin hey! Thank you for your answer and demo :) . The thing is, once you set `setIgnoreMouseEvents` to true, the electron window doesn't detect mouse clicks at all ( kind of what you expect with `setIgnoreMouseEvents`). I tried capturing them in the same way as you demonstrated, but unfortunately the mouse click event doesn't even get into the `BrowserWindow`, it just gets through to whatever is behind the overlay.

**Comment by nikkwong** [comment]
Getting this to work made me want to instantly give up on this lib. Total waste of time and very frustrating, especially with `-webkit-app-region` being ignored with `setIgnoreMouseEvents`... what a mess. Fortunately I have a solution that is working for me. 

Make sure to enable `webPreferences.nodeIntegration`, then in your HTML: 

Css
```
body {pointer-events: none}
* {pointer-events: all;}
.draggable {-webkit-app-region: drag;}
```

Js
```
setIgnoreMouseEvents = window.require('electron').remote.getCurrentWindow().setIgnoreMouseEvents

window.addEventListener('mousemove', event => {
  if (event.target === document.documentElement) {
    window.setIgnoreMouseEvents(true, {forward: true})   // {forward: true} keeps generating MouseEvents
    if (t) clearTimeout(t)
    t = setTimeout(function() {
      window.setIgnoreMouseEvents(false)
    }, 150)
  } else window.setIgnoreMouseEvents(false)
})
```
The idea here is that you by default don't want to ignore mouse events. But, if the mouse ever moves over the window itself, we want to ignore that event and pass-through it. This does not include any children of the body, which WILL receive clicks, and won't pass-through.

AKA body and HTML event targets => pass-through
Children of body => don't pass through

The timeout is necessary because if you `mousemove` from a pass-through element (like body or html) to a `.draggable` element (i.e. has `-webkit-app-region` applied), then it won't receive click events, and `setIgnoreMouseEvents` will never be set back to `false`. So, we try to just set it to `false` by default (after timeout).

**Comment by vifird** [comment]
> @singuerinc yes, if you enable `disableHardwareAcceleration` and set the app transparency to true, only solid and `if you set the opacity to 0.1` will be clickable. Do take note about the slowdown on macs (check @adeperio's response)

This works on Windows10, but not on MacOSX.

**Comment by alectrocute** [comment]
Hey people. What's the agreed upon best practice for this nowadays, if any?

**Comment by DevelopSmith** [comment]
@nikkwong this is the only solution to work for me on both MacOS and Windows. You made my day!

**Comment by rbarisic-lme** [comment]
@nikkwong, this causes the mouse to behave weirdly if interacting with the window behind the electron window's transparent area

**Comment by alexandreroche** [comment]
It would be great if there were just a way to set a callback that returns true or false depending on whether you want to window to respond to clicks. On Mac, this could be implemented by conditionally return nil or self from https://github.com/electron/electron/blob/master/shell/browser/native_browser_view_mac.mm#L36.

**Comment by dingchaoyan1983** [comment]
because the transparent window is lost focus, so the 
```
window.addEventListener('mousemove', cb)
```
can not triggered

**Comment by DaniGuardiola** [comment]
Tested @MickeyPickles fix, works on Linux but messes with mouse events a lot.

Currently working on a decent Linux workaround using [screen.getCursorScreenPoint()
](https://www.electronjs.org/docs/api/screen#screengetcursorscreenpoint) and [Document.elementFromPoint()](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint) somehow with IPC between bg and renderer processes.

I Will post it here if I get it to work properly.

**Comment by joepie91** [comment]
@DaniGuardiola Did you make any further progress on that?

**Comment by DaniGuardiola** [comment]
@joepie91 the project I was working on is on hold because I just started at a new job. If/when I tackle this, I'll do my best to report back to this issue.

**Comment by DDinenno** [comment]
@nikkwong Your example was the only one that **mostly** worked.  

The problem is that, if you rest you mouse for more than the timeout amount, you are unable to click through.  

I made another example (See below), which is a modified version of yours.  
What this achieves, is it makes sure that the element is not the document because setting ignore to false.
It's done by adding a interval which inside will get the current screen position and using **document.elementFromPoint**, it will grab the top most element, if there's no elements it will return the **documentElement**.


main.js 
`
ipcMain.handle("get-screen-point", () => screen.getCursorScreenPoint());
`

preload.js added a bridge function
`
 getScreenPoint: () => ipcRenderer.invoke("get-screen-point"),
`

handle mouse move
`
 const handleMouseMove = (e) => {
      if (e.target === menuWindow.current.document.documentElement) {
        window.api.menuOverlaySetIgnoreMouseEvents(true);

        if (t) clearInterval(t);
        t = setInterval(async () => {
          const point = window.api.getScreenPoint();

          const elem = menuWindow.current.document.elementFromPoint(
            Math.floor(point.x),
            Math.floor(point.y)
          );

          if (elem !== menuWindow.current.document.documentElement) {
            window.api.menuOverlaySetIgnoreMouseEvents(false);
            clearInterval(t);
          }
        }, 150);
      } else {
        window.api.menuOverlaySetIgnoreMouseEvents(false);
      }
    };
`

**Comment by whosnorman** [comment]
I was able to use @LauraAubin's approach with a couple tweaks! 

* Instead of using `mainWindow.hide()` I use `Menu.sendActionToFirstResponder('hide:')`. I [found it here](https://stackoverflow.com/a/60941280) cause when I was trying to CMD+Tab back to the app there was no event firing to be able to use `mainWindow.show()`. This approach also gives up control to the application you click on below your app, so no need for double clicking. You can get the Menu variable via `const { ..., ipcMain, Menu } = require('electron');` 

* Also, instead of adding an id to all elements that should be clicked, I just listen for if the event target is the document body `e.target == document.body`, and put the rest of my app in an element that has a mask and its overflow is hidden.

**Comment by aarowman** [comment]
Any updates on this?

I'd also be interested to see the `hover` behavior go through transparency (in addition to clicks).

My use case is to overlay a game, and the only way the click through would work is by disabling GPU compositing or acceleration, but the click/hover goes through everything - even non-transparent elements.

**Comment by LZQCN** [comment]
> During the development of my Windows gadget project which requires click-thru, I have unintentionally come up with a somewhat fancy but practical solution which works to define a clickcable area of truly arbitrary shape.
> 
> In the renderer process, use "screen.getCursorScreenPoint()" to keep tracking the cursor position, and then send the coordinates to the main process regularly. On the main process, use "webContents.capturePage()" to capture the image of that exact 1x1 pixel on the renderer window. Since it's just 1 pixel, the performance cost is ignorable, and then you can define whether that pixel is clickable or not by looking at the opacity value. This way, you can practically define a truly pixel-correct clickable area. The only issue left here is a function to detect the mousedown state regardless of the cursor position.

We always stand on the shoulders of giants.
Combining the ideas of geniuses, the following code achieves perfect results.

```typescript
  setInterval(() => {
    const point = screen.getCursorScreenPoint();
    const [x, y] = win.getPosition();
    const [w, h] = win.getSize();

    if (point.x > x && point.x < x + w && point.y > y && point.y < y + h) {
      updateIgnoreMouseEvents(point.x - x, point.y - y);
    }
  }, 300);

  const updateIgnoreMouseEvents = async (x, y) => {
    console.log("updateIgnore");

    // capture 1x1 image of mouse position.
    const image = await win.webContents.capturePage({
      x,
      y,
      width: 1,
      height: 1,
    });

    var buffer = image.getBitmap();

    // set ignore mouse events by alpha.
    win.setIgnoreMouseEvents(!Boolean(buffer[3]));
    console.log("setIgnoreMouseEvents", !Boolean(buffer[3]));
  };
```

**Comment by raphael10-collab** [comment]
@LZQCN I tried to follow your suggestion here: https://github.com/electron/electron/issues/1335#issuecomment-1585787243 but , but I guess I'm making something wrong because nothing happens:

`SearchWindow` is in my case the `main window` . 

declare const SEARCH_WINDOW_PRELOAD_WEBPACK_ENTRY: string;

let SearchWindow: BrowserWindow

let child_window: BrowserWindow
declare const CHILD_WINDOW_WEBPACK_ENTRY: string;
declare const CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY: string;

const createSearchWindow = (): void => {
    console.log("createSearchWindow called")

    SearchWindow = new BrowserWindow({
        fullscreen: false,
        height: 900,
        width: 1200,
        webPreferences: {
            preload: SEARCH_WINDOW_PRELOAD_WEBPACK_ENTRY,
            nodeIntegration: false,
            contextIsolation: true,
            nodeIntegrationInWorker: false,
            nodeIntegrationInSubFrames: false,
            webviewTag: true,
            webSecurity: true
        }
    });

    SearchWindow.loadURL(SEARCH_WINDOW_WEBPACK_ENTRY)

    // Open the DevTools
    if (isDev) {
        SearchWindow.webContents.openDevTools()
    }
    SearchWindow.webContents.on("did-navigate", async (event, url) => {
        SearchWindow.webContents.session.cookies.get({name:"_session_id"})
    })

    let searchWindowHeight: number = SearchWindow.getBounds().height
    let searchWindowWidth: number = SearchWindow.getBounds().width
    let searchWindow_x: number = SearchWindow.getBounds().x
    let searchWindow_y: number = SearchWindow.getBounds().y

    child_window = new BrowserWindow({
        //height: 900,
        //width: 1200,
        height: searchWindowHeight,
        width: searchWindowWidth,
        x: searchWindow_x,
        y: searchWindow_y,
        maximizable: false,
        minimizable: false,
        parent: SearchWindow,
        transparent: true,
        skipTaskbar: true,
        movable: false,
        resizable: false,
        alwaysOnTop: true,
        webPreferences: {
            nodeIntegration: false,
            nodeIntegrationInSubFrames: true,
            scrollBounce: true,
            safeDialogs: true,
            safeDialogsMessage: 'Prevent this page from creating additional dialogs',
            preload: CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY,
            contextIsolation: true,
            sandbox: true,
            enableWebSQL: false,
            // match Chrome's default for anti-fingerprinting purposes (Electron defaults to 0)
            minimumFontSize: 6
        }
    })

    child_window.once('ready-to-show', () => {
        child_window.show()
    })

    child_window.setFocusable(false)

    child_window.setIgnoreMouseEvents(true,{
        forward: true ,
    })

    setInterval(() => {
        const point = screen.getCursorScreenPoint()
        const [x, y] = SearchWindow.getPosition()
        const [w, h] = SearchWindow.getSize()

        if (point.x > x && point.x < x + w && point.y > y && point.y < y + h) {
            updateIgnoreMouseEvents(point.x - x, point.y - y)
        }
    }, 300)

    const updateIgnoreMouseEvents = async (x, y) => {
        console.log("updateIgnoreMouseEvents")

        // capture 1x1 image of mouse position.
        const image = await SearchWindow.webContents.capturePage({
            x,
            y,
            width: 1,
            height: 1,
        })

        let buffer = image.getBitmap()

        // set ignore mouse events by alpha.
        SearchWindow.setIgnoreMouseEvents(!buffer[3])
        console.log("setIgnoreMouseEvents", !buffer[3])
    }


How to make it work ?

**Comment by LZQCN** [comment]
> @LZQCN I tried to follow your suggestion here: [#1335 (comment)](https://github.com/electron/electron/issues/1335#issuecomment-1585787243) but , but I guess I'm making something wrong because nothing happens:
> 
> `SearchWindow` is in my case the `main window` .
> 
> declare const SEARCH_WINDOW_PRELOAD_WEBPACK_ENTRY: string;
> 
> let SearchWindow: BrowserWindow
> 
> let child_window: BrowserWindow declare const CHILD_WINDOW_WEBPACK_ENTRY: string; declare const CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY: string;
> 
> const createSearchWindow = (): void => { console.log("createSearchWindow called")
> 
> ```
> SearchWindow = new BrowserWindow({
>     fullscreen: false,
>     height: 900,
>     width: 1200,
>     webPreferences: {
>         preload: SEARCH_WINDOW_PRELOAD_WEBPACK_ENTRY,
>         nodeIntegration: false,
>         contextIsolation: true,
>         nodeIntegrationInWorker: false,
>         nodeIntegrationInSubFrames: false,
>         webviewTag: true,
>         webSecurity: true
>     }
> });
> 
> SearchWindow.loadURL(SEARCH_WINDOW_WEBPACK_ENTRY)
> 
> // Open the DevTools
> if (isDev) {
>     SearchWindow.webContents.openDevTools()
> }
> SearchWindow.webContents.on("did-navigate", async (event, url) => {
>     SearchWindow.webContents.session.cookies.get({name:"_session_id"})
> })
> 
> let searchWindowHeight: number = SearchWindow.getBounds().height
> let searchWindowWidth: number = SearchWindow.getBounds().width
> let searchWindow_x: number = SearchWindow.getBounds().x
> let searchWindow_y: number = SearchWindow.getBounds().y
> 
> child_window = new BrowserWindow({
>     //height: 900,
>     //width: 1200,
>     height: searchWindowHeight,
>     width: searchWindowWidth,
>     x: searchWindow_x,
>     y: searchWindow_y,
>     maximizable: false,
>     minimizable: false,
>     parent: SearchWindow,
>     transparent: true,
>     skipTaskbar: true,
>     movable: false,
>     resizable: false,
>     alwaysOnTop: true,
>     webPreferences: {
>         nodeIntegration: false,
>         nodeIntegrationInSubFrames: true,
>         scrollBounce: true,
>         safeDialogs: true,
>         safeDialogsMessage: 'Prevent this page from creating additional dialogs',
>         preload: CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY,
>         contextIsolation: true,
>         sandbox: true,
>         enableWebSQL: false,
>         // match Chrome's default for anti-fingerprinting purposes (Electron defaults to 0)
>         minimumFontSize: 6
>     }
> })
> 
> child_window.once('ready-to-show', () => {
>     child_window.show()
> })
> 
> child_window.setFocusable(false)
> 
> child_window.setIgnoreMouseEvents(true,{
>     forward: true ,
> })
> 
> setInterval(() => {
>     const point = screen.getCursorScreenPoint()
>     const [x, y] = SearchWindow.getPosition()
>     const [w, h] = SearchWindow.getSize()
> 
>     if (point.x > x && point.x < x + w && point.y > y && point.y < y + h) {
>         updateIgnoreMouseEvents(point.x - x, point.y - y)
>     }
> }, 300)
> 
> const updateIgnoreMouseEvents = async (x, y) => {
>     console.log("updateIgnoreMouseEvents")
> 
>     // capture 1x1 image of mouse position.
>     const image = await SearchWindow.webContents.capturePage({
>         x,
>         y,
>         width: 1,
>         height: 1,
>     })
> 
>     let buffer = image.getBitmap()
> 
>     // set ignore mouse events by alpha.
>     SearchWindow.setIgnoreMouseEvents(!buffer[3])
>     console.log("setIgnoreMouseEvents", !buffer[3])
> }
> ```
> 
> How to make it work ?

I found that in code `SearchWindow` does not seem to be a transparent window, but `child_window` is a transparent window, And the `updateIgnoreMouseEvents` function is updating the `ignore mouse events` state of `SearchWindow`, not `child_window`.

To make it work for `child_window`, you need to update the line `SearchWindow.setIgnoreMouseEvents(!buffer[3])` to `child_window.setIgnoreMouseEvents(!buffer[3])`.

**Comment by raphael10-collab** [comment]
Yes. the transparent window is `child_window`

I've put in your code the right window : 

    setInterval(() => {
        const point = screen.getCursorScreenPoint()
        const [x, y] = child_window.getPosition()
        const [w, h] = child_window.getSize()

        if (point.x > x && point.x < x + w && point.y > y && point.y < y + h) {
            updateIgnoreMouseEvents(point.x - x, point.y - y)
        }
    }, 300)

    const updateIgnoreMouseEvents = async (x, y) => {
        console.log("updateIgnoreMouseEvents")

        // capture 1x1 image of mouse position.
        const image = await child_window.webContents.capturePage({
            x,
            y,
            width: 1,
            height: 1,
        })

        let buffer = image.getBitmap()

        // set ignore mouse events by alpha.
        child_window.setIgnoreMouseEvents(!buffer[3])
        console.log("setIgnoreMouseEvents", !buffer[3])
    }

The`child_window` (the transparent one) is on top of the `SearchWindow`.
Clicking on the `child_window`, the one which is on top, does not pass the click event to the underneath window 

![image](https://github.com/electron/electron/assets/70313067/8d090a5c-4791-4d6d-b79c-37c45149e99a)


and I'm getting this error : 

![image](https://github.com/electron/electron/assets/70313067/61e8346d-3441-449b-a0e8-2619589ad034)

![image](https://github.com/electron/electron/assets/70313067/5714fa25-8165-4f83-9e21-974c501c0c4d)

The error message points to this line: 

    setInterval(() => {
        const point = electron_1.screen.getCursorScreenPoint();
        const [x, y] = child_window.getPosition();    // <----------------------------------------------------------------------------
        const [w, h] = child_window.getSize();
        if (point.x > x && point.x < x + w && point.y > y && point.y < y + h) {
            updateIgnoreMouseEvents(point.x - x, point.y - y);
        }
    }, 300);

**Comment by LZQCN** [comment]
@raphael10-collab 

```
const timer = setInterval(() => {
    const point = electron_1.screen.getCursorScreenPoint();
    const [x, y] = child_window.getPosition();    // <----------------------------------------------------------------------------
    const [w, h] = child_window.getSize();
    if (point.x > x && point.x < x + w && point.y > y && point.y < y + h) {
        updateIgnoreMouseEvents(point.x - x, point.y - y);
    }
}, 300);

child_window.on("close", () => {
    clearInterval(timer);
});
```

**Comment by raphael10-collab** [comment]
@LZQCN With `clearInterval(timer)` the `Timeout Error` disappeared 

This is the output 

![image](https://github.com/electron/electron/assets/70313067/11304727-06b1-4823-8651-e68a4511f443)

The `transparent` window, the `child_window` does not allow me to click the button `Click me` built in the underneath window  (`SearchWindow`): how to "`click-through`" the trasparent top window, in order to be able to click the button in the underneath window? Is that possible?

**Comment by raphael10-collab** [comment]
Following the indications found here: https://stackoverflow.com/questions/53357428/how-to-make-mouse-pass-through-window-in-electron I've put `focusable: false` for the `child_window` (the `transparent window) : 

    child_window = new BrowserWindow({
        //height: 900,
        //width: 1200,
        height: searchWindowHeight,
        width: searchWindowWidth,
        x: searchWindow_x,
        y: searchWindow_y,
        maximizable: false,
        minimizable: false,
        parent: SearchWindow,
        transparent: true,
        skipTaskbar: true,
        movable: false,
        resizable: false,
        focusable: false, // https://stackoverflow.com/questions/53357428/how-to-make-mouse-pass-through-window-in-electron
        frame: false,
        alwaysOnTop: true,
        webPreferences: {
            nodeIntegration: false,
            nodeIntegrationInSubFrames: true,
            scrollBounce: true,
            safeDialogs: true,
            safeDialogsMessage: 'Prevent this page from creating additional dialogs',
            preload: CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY,
            contextIsolation: true,
            sandbox: true,
            enableWebSQL: false,
            // match Chrome's default for anti-fingerprinting purposes (Electron defaults to 0)
            minimumFontSize: 6
        }
    })

Now the top `transparent` window is not visible anymore, the button `Click me` in the underneath window is clickable and the `screen.getCursorScreenPoint()` yields its output

![image](https://github.com/electron/electron/assets/70313067/6216b6e1-60b5-492d-907b-8d876c67382e)

But, still, I'm not able to get the correct mouse position with `getCursorScreenPoint()`.... I mean... what I would like to do is to get, in the `transparent` window, `child_window`, the position `{x, y}` of the  mouse when clicking on the underneath window

**Comment by LZQCN** [comment]
@raphael10-collab My code works for this scenario:

https://github.com/electron/electron/assets/43063927/e29e9bf6-748e-42d1-8efd-e6c60d1b46ed

**Comment by LZQCN** [comment]
My code works for this scenario:
![QQ录屏20230710002937](https://github.com/electron/electron/assets/43063927/14b82c2d-b867-4e27-a97f-bb2f6170351a)

**Comment by raphael10-collab** [comment]
Wow ......  I do not understand how you are actually using the `screen.getCursorScreenPoint` 's ouput:
How are you passing the cursor position from the `transparent` window to the underneath window ?

**Comment by LZQCN** [comment]
If I use my code, clicking on the transparent area of the window no longer generates any mouse events for the window. To be precise, the transparent window will be set to ignore any mouse events when the mouse is over the transparent area. only the content below the transparent window will be clicked.

**Comment by LZQCN** [comment]
@raphael10-collab 
If I use my code, clicking on the transparent area of the window no longer generates any mouse events for the window. To be precise, the transparent window will be set to ignore any mouse events when the mouse is over the transparent area. only the content below the transparent window will be clicked.

**Comment by LZQCN** [comment]
@raphael10-collab 
If use my code, clicking on the transparent area of the window no longer generates any mouse events for the window. To be precise, the transparent window will be set to ignore any mouse events when the mouse is over the transparent area. only the content below the transparent window will be clicked.

**Comment by raphael10-collab** [comment]
@LZQCN I've set : 

    child_window.setFocusable(false)

    child_window.setIgnoreMouseEvents(true,{
        forward: true ,
    })

Is this not enough ?

**Comment by raphael10-collab** [comment]
I do not understand why this message sent from the `child_window` (the `transparent` window on top of the other window) is not received in the main process:  

 `child_window` renderer : 

  React.useEffect(() => {
        window.api.electronIpcSend("hello-from-app-child", "HELLOOOOOOOOO")

    }, [])

in `main` : 

    ipcMain.on("hello-from-app-child", (event, args) => {
        console.log("IpcMain.on-hello-from-app-child-args: ", args)
    })

This is how I define and create the bottom_window (SearchWindow) and the trasparent window on top  :  

declare const SEARCH_WINDOW_WEBPACK_ENTRY: string;
declare const SEARCH_WINDOW_PRELOAD_WEBPACK_ENTRY: string;

let SearchWindow: BrowserWindow

let child_window: BrowserWindow
declare const CHILD_WINDOW_WEBPACK_ENTRY: string;
declare const CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY: string;

const createSearchWindow = (): void => {
    console.log("createSearchWindow called")

    SearchWindow = new BrowserWindow({
        fullscreenable: false,
        height: 900,
        width: 1200,
        kiosk: true,
        webPreferences: {
            preload: SEARCH_WINDOW_PRELOAD_WEBPACK_ENTRY,
            nodeIntegration: false,
            contextIsolation: true,
            nodeIntegrationInWorker: false,
            nodeIntegrationInSubFrames: false,
            webviewTag: true,
            webSecurity: true
        }
    let searchWindowHeight: number = SearchWindow.getBounds().height
    let searchWindowWidth: number = SearchWindow.getBounds().width
    let searchWindow_x: number = SearchWindow.getBounds().x
    let searchWindow_y: number = SearchWindow.getBounds().y

    child_window = new BrowserWindow({
        height: searchWindowHeight,
        width: searchWindowWidth,
        fullscreenable: false,
        x: searchWindow_x,
        y: searchWindow_y,
        maximizable: false,
        minimizable: false,
        parent: SearchWindow,
        transparent: true,
        skipTaskbar: true,
        movable: false,
        resizable: false,
        //focusable: false, 
        frame: false,
        alwaysOnTop: true,
        webPreferences: {
            nodeIntegration: false,
            nodeIntegrationInSubFrames: true,
            scrollBounce: true,
            safeDialogs: true,
            safeDialogsMessage: 'Prevent this page from creating additional dialogs',
            preload: CHILD_WINDOW_PRELOAD_WEBPACK_ENTRY,
            contextIsolation: true,
            sandbox: true,
            enableWebSQL: false,
            // match Chrome's default for anti-fingerprinting purposes (Electron defaults to 0)
            minimumFontSize: 6
        }
    })

    child_window.once('ready-to-show', () => {
        child_window.show()
    })

    child_window.setIgnoreMouseEvents(true,{
        forward: true ,
    })

}

**Comment by foxxy432** [comment]
It is ignored by Microsoft because of incompatibilities BUT you can use my fork: https://github.com/lnxx-kernel/electron_touch

**Comment by SevenFantastic** [comment]
I've noticed that the Electron API setShape can solve some problems, similar to the Windows setWindowRgn. It allows areas without a set shape to be non-drawable and lets mouse events pass through. 
![image](https://github.com/electron/electron/assets/20679462/4dea00e2-6128-4699-a5d1-7025041b4052)
This solved my problem. If your page's displayed content is not complex, you can use this API.

**Comment by Nebula-Developer** [comment]
[win.SetShape](https://www.electronjs.org/docs/latest/api/browser-window#winsetshaperects-windows-linux-experimental) is currently only supported (experimentally) on Windows and Linux. Are there any efforts towards implementing this on macOS? (and removing it from its 'experimental' state)

**Comment by tisbutanalt2** [comment]
Referring to the workaround proposed by @LZQCN [#1355 (comment)](https://github.com/electron/electron/issues/1335#issuecomment-1585787243) I suggest adding an additional check if the window won't have any moving elements.

- Create variables for the last checked X and Y positions
- Return if the current X and Y positions are the same

**Note:** This is mostly useful only if your window does NOT have any moving elements.

**Comment by shady-xia** [comment]
After setting `transparent` on Windows, click-through functionality is enabled, but this currently does not work on Linux.
My solution involves using `setShape`: by monitoring DOM changes via a MutationObserver, I retrieve the dimensions of each visible element, assemble them into a rect array representing the window, and then invoke `setShape`.

**Comment by deadcoder0904** [comment]
Would love to have this. In the meantime, this seems useful (https://github.com/loomhq/ElectronMacOSClickThrough)

**Comment by BryanEnid** [comment]
I found a workaround.

I have defined a global variable to toggle between "Interactive" or "click-through" in my front end. I set this value based on mouse events like onMouseEnter onMouseLeave, etc..

which is then calling on the electron side this:
```js
export class ModeManager implements AppModule {
  enable({ app }: ModuleContext): void {
    ipcMain.on("set-mode", async (_, mode) => {
      const win = BrowserWindow.getAllWindows()[0];

      if (mode === "interactive") win.setIgnoreMouseEvents(false);
      if (mode === "click-through") win.setIgnoreMouseEvents(true, { forward: true });
    });
  }
}
```

In this way, I can click through transparent backgrounds and when the mouse is on top of an element it will make it interactive once more.

**Comment by Eneris** [comment]
> I found a workaround.
> 
> I have defined a global variable to toggle between "Interactive" or "click-through" in my front end. I set this value based on mouse events like onMouseEnter onMouseLeave, etc..
> 
> which is then calling on the electron side this:
> 
> export class ModeManager implements AppModule {
>   enable({ app }: ModuleContext): void {
>     ipcMain.on("set-mode", async (_, mode) => {
>       const win = BrowserWindow.getAllWindows()[0];
> 
>       if (mode === "interactive") win.setIgnoreMouseEvents(false);
>       if (mode === "click-through") win.setIgnoreMouseEvents(true, { forward: true });
>     });
>   }
> }
> In this way, I can click through transparent backgrounds and when the mouse is on top of an element it will make it interactive once more.

@BryanEnid Watch out for the bug on Windows when the cursor glitches and blinks while moving it over anything that changes the cursor shape when passing the mouse events
